test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/SetBasedOperationTests.cs

```diff
@@ -0,0 +1,420 @@
+        {
+            ParseUriAndVerify(
+                new Uri("http://gobbledygook/People/$filter(@p1)/$each/Fully.Qualified.Namespace.SummonPuppies?@p1=ID eq 1"),
+                (oDataPath, filterClause, aliasNodes) =>
+                {
+                    oDataPath.Count.Should().Be(4);
+
+                    List<FilterSegment> filterSegments = oDataPath.OfType<FilterSegment>().ToList();
+                    filterSegments.Count.Should().Be(1);
+                    filterSegments[0].Expression.ShouldBeParameterAliasNode("@p1", EdmCoreModel.Instance.GetBoolean(false));
+                    filterSegments[0].TargetEdmType.ToString().ShouldBeEquivalentTo(HardCodedTestModel.GetPersonType().ToString());
+                    filterSegments[0].SingleResult.Should().BeFalse();
+
+                    filterClause.Should().BeNull();
+                    aliasNodes["@p1"].ShouldBeBinaryOperatorNode(BinaryOperatorKind.Equal).And.Right.ShouldBeConstantQueryNode(1);
+
+                    List<EachSegment> eachSegments = oDataPath.OfType<EachSegment>().ToList();
+                    eachSegments.Count.Should().Be(1);
+                    eachSegments[0].TargetEdmNavigationSource.Should().Be(HardCodedTestModel.GetPeopleSet());
+                    eachSegments[0].TargetEdmType.Should().Be(HardCodedTestModel.GetPersonType());
+                });
+        }
+
+        // NOTE: Per OData 4.01 spec, GET operation and functions may proceed $each, but we are limiting the scope of that feature
+        // by permitting only ONE action segment to follow $each.
+        [Fact]
+        public void SetBasedOperations_EachSegmentOnSetBasedActionOnCollectionResults_ThrowsException()
+        {
+            Action parse = () => ParseUriAndVerify(
+                new Uri("http://gobbledygook/People/$each/Fully.Qualified.Namespace.SummonPuppies/$each"),
+                (oDataPath, filterClause, aliasNodes) =>
+                {
+                });
+            parse.ShouldThrow<ODataException>().WithMessage(ODataErrorStrings.RequestUriProcessor_MustBeLeafSegment("Fully.Qualified.Namespace.SummonPuppies"));
+        }
+
+        // NOTE: Per OData 4.01 spec, GET operation and functions may proceed $each, but we are limiting the scope of that feature
+        // by permitting only ONE action segment to follow $each.
+        [Fact]
+        public void SetBasedOperations_FunctionAfterEachSegment_ThrowsException()
+        {
+            Action parse = () => ParseUriAndVerify(
+                new Uri("http://gobbledygook/People/$each/Fully.Qualified.Namespace.AllMyFriendsDogs"),
+                (oDataPath, filterClause, aliasNodes) =>
+                {
+                });
+            parse.ShouldThrow<ODataException>().WithMessage(ODataErrorStrings.RequestUriProcessor_OnlySingleActionCanProceedEachPathSegment);
+        }
+
+        // NOTE: Per OData 4.01 spec, GET operation and functions may proceed $each, but we are limiting the scope of that feature

```

Comment:
>proceed [](start = 70, length = 7)

follow?

Reply:
:+1:

Reply:
Not fixed?  Also, wording is a bit awkward; a function is an operations, but we are checking for actions, which are also operations. I think what we are really enforcing is only one operation segment (function or action) can follow /$each? (is that right, that a function can follow $each?)

---
In reply to: [224938789](https://github.com/OData/odata.net/pull/1110#discussion_r224938789) [](ancestors = 224938789)

Reply:
Distinctly remember updating the text. Marking unresolved.

Reply:
I think we decided that only one action can follow $each.

For convenience, here's the [spec](http://docs.oasis-open.org/odata/odata/v4.01/cs01/part1-protocol/odata-v4.01-cs01-part1-protocol.html#sec_BindinganOperationtoaResource).